### PR TITLE
[SkyServe] Serving with Spot

### PR DESCRIPTION
**Pull request by MaoZiming**
#2596 rebase

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [x] Smoke test: `pytest tests/test_smoke.py::test_skyserve_spot_user_bug` 
- [x] Smoke test: `pytest tests/test_smoke.py::test_skyserve_spot_recovery` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/serve/replica_managers.py [hunk 1]
```diff
@@ -490,6 +503,15 @@ def scale_down(self, replica_ids: List[int]) -> None:
         for replica_id in replica_ids:
             self._terminate_replica(replica_id, sync_down_logs=False)
 
+    def _recover_from_preemption(self, replica_id: int) -> None:
```
Comment (cblmemo) [inline sky/serve/replica_managers.py:506]: Add a TODO for possible recover strategy?

File: sky/serve/replica_managers.py [hunk 2]
```diff
@@ -673,6 +700,28 @@ def _probe_all_replicas(self) -> None:
                 if info.status_property.first_ready_time is None:
                     info.status_property.first_ready_time = probe_time
             else:
+                if self.use_spot:
```
Comment (cblmemo) [inline sky/serve/replica_managers.py:703]:
```suggestion
            else:
                handle = global_user_state.get_handle_from_cluster_name(info.cluster_name)
                if handle.launched_resources.use_spot:
```
Shall we read from the `handle.launched_resources` here?

File: sky/serve/serve_state.py [hunk 3]
```diff
@@ -81,6 +81,9 @@ class ReplicaStatus(enum.Enum):
     # we should guarantee no resource leakage like regular sky.
     FAILED_CLEANUP = 'FAILED_CLEANUP'
 
+    # The replica spot VM is preempted by the cloud provider.
```
Comment (cblmemo) [inline sky/serve/serve_state.py:84]:
```suggestion
    # The replica is a spot VM and it is preempted by the cloud provider.
```

File: sky/utils/cli_utils/status_utils.py [hunk 4]
```diff
@@ -491,14 +491,14 @@ def _get_replica_resources(replica_record: _ReplicaRecord) -> str:
         return '-'
     assert isinstance(handle, backends.CloudVmRayResourceHandle)
     cloud = handle.launched_resources.cloud
-    launched_resource_str = f'{cloud}'
     if handle.launched_resources.accelerators is None:
         vcpu, _ = cloud.get_vcpus_mem_from_instance_type(
             handle.launched_resources.instance_type)
-        launched_resource_str += f'(vCPU={int(vcpu)})'
+        hardware = f'vCPU={int(vcpu)})'
```
Comment (cblmemo) [inline sky/utils/cli_utils/status_utils.py:497]:
```suggestion
        hardware = f'vCPU={int(vcpu)}'
```
nit

File: sky/utils/cli_utils/status_utils.py [hunk 5]
```diff
@@ -491,14 +491,14 @@ def _get_replica_resources(replica_record: _ReplicaRecord) -> str:
...
     else:
-        launched_resource_str += f'({handle.launched_resources.accelerators})'
-    resources_str = (f'{handle.launched_nodes}x {launched_resource_str}')
+        hardware = f'{handle.launched_resources.accelerators})'
+    spot = '[Spot]' if handle.launched_resources.use_spot else ''
+    resources_str = f'{handle.launched_nodes}x {cloud}({hardware}{spot})'
```
Comment (cblmemo) [inline sky/utils/cli_utils/status_utils.py:501]:
```suggestion
    resources_str = f'{handle.launched_nodes}x {cloud}({spot}{hardware})'
```
nit: align with `sky.Resources.__repr__`

File: tests/skyserve/spot/non_recovery.py
Comment (cblmemo) [inline tests/skyserve/spot/non_recovery.py:1]: nit: EOF

File: tests/test_smoke.py [hunk 6]
```diff
@@ -2734,6 +2734,62 @@ def generate_llm_test_command(prompt: str, expected_output: str) -> str:
...
+        query_cmd = (f'gcloud compute instances list --filter='
+                     f'"(labels.ray-cluster-name:{cluster_name})" '
+                     f'--zones={zone} --format="value(name)"')
+        return (f'gcloud compute instances delete --zone={zone}'
+                f' --quiet $({query_cmd})')
+
+    test = Test(
+        f'test-skyserve-spot-recovery-gcp',
+        [
+            f'sky serve up -n {name} -y tests/skyserve/spot_recovery/gcp.yaml',
```
Comment (cblmemo) [inline tests/test_smoke.py:2755]: Seems like the file path is incorrect?

File: tests/test_smoke.py [hunk 7]
```diff
@@ -2734,6 +2734,62 @@ def generate_llm_test_command(prompt: str, expected_output: str) -> str:
...
+
+@pytest.mark.gcp
+@pytest.mark.sky_serve
+def test_skyserve_spot_nonrecovery():
+    """Tests that spot recovery doesn't occur for non-preemption failures"""
+    name = _get_service_name()
+    test = Test(
+        f'test-skyserve-spot-nonrecovery-gcp',
+        [
+            f'sky serve up -n {name} -y tests/skyserve/spot_recovery/nonrecovery.yaml',
```
Comment (cblmemo) [inline tests/test_smoke.py:2776]: ditto

**Review by cblmemo — COMMENTED** [review]
Thanks for the prompt rebase!! Left several comments 🫡

File: sky/serve/replica_managers.py (hunk 2 above)
Comment (MaoZiming) [reply sky/serve/replica_managers.py:703]: yep that should work!

File: sky/serve/replica_managers.py [hunk 8]
```diff
@@ -169,6 +170,8 @@ class ReplicaStatusProperty:
     first_ready_time: Optional[float] = None
     # None means sky.down is not called yet.
     sky_down_status: Optional[ProcessStatus] = None
+    # The replica's spot instance was preempted.
+    preempted: bool = False
 
     def is_scale_down_succeeded(self, initial_delay_seconds: int,
```
[the remark this answers is not in the document]
Comment (MaoZiming) [reply sky/serve/replica_managers.py:176]: Is it under `should_track_status`?

[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/serve/replica_managers.py:176]: Then what does this comment mean..?
image

[the remark this answers is not in the document]
Comment (MaoZiming) [reply sky/serve/replica_managers.py:176]: ah I see. Thanks

**Comment by MaoZiming** [comment]
@cblmemo Thanks! Just addressed the comments 🫡

File: sky/serve/replica_managers.py [hunk 9]
```diff
@@ -195,6 +198,8 @@ def is_scale_down_succeeded(self, initial_delay_seconds: int,
             return False
         if not self.service_ready_now:
             return False
+        if self.preempted:
+            return True
```
Comment (cblmemo) [inline sky/serve/replica_managers.py:202]:
```suggestion
        if self.preempted:
            return True
        if not self.service_ready_now:
            return False
```
If the `preempted` is set to `True`, then IIUC it is always after a probe failure and thus the `self.service_ready_now` must be `True`?

File: sky/serve/replica_managers.py [hunk 10]
```diff
@@ -204,13 +209,17 @@ def should_track_status(self) -> bool:
...
         return True
 
     def to_replica_status(self) -> serve_state.ReplicaStatus:
         if self.sky_launch_status == ProcessStatus.RUNNING:
             # Still launching
             return serve_state.ReplicaStatus.PROVISIONING
         if self.sky_down_status is not None:
+            if self.preempted:
```
Comment (cblmemo) [inline sky/serve/replica_managers.py:221]: nit: Add a comment

File: sky/resources.py
Comment (cblmemo) [inline sky/resources.py:1]: Do we still need changes in this file?

File: sky/serve/replica_managers.py [hunk 11]
```diff
@@ -257,12 +266,14 @@ def to_replica_status(self) -> serve_state.ReplicaStatus:
...
 
-    def __init__(self, replica_id: int, cluster_name: str) -> None:
+    def __init__(self, replica_id: int, cluster_name: str,
+                 use_spot: bool) -> None:
         self.replica_id: int = replica_id
         self.cluster_name: str = cluster_name
         self.first_not_ready_time: Optional[float] = None
         self.consecutive_failure_times: List[float] = []
         self.status_property: ReplicaStatusProperty = ReplicaStatusProperty()
+        self.use_spot: bool = use_spot
```
Comment (cblmemo) [inline sky/serve/replica_managers.py:276]: Since now we are peeking at the `launched_resources`, do we still need these `use_spot`s?

**Review by cblmemo — COMMENTED** [review]
Thanks for the prompt fix! Left several comments, mainly nits and redundant `use_spot`s 🚀

File: sky/resources.py
Comment (MaoZiming) [reply sky/resources.py:1]: let me remove them

File: sky/serve/replica_managers.py (hunk 11 above)
Comment (MaoZiming) [reply sky/serve/replica_managers.py:276]: ditto

Comment (MaoZiming) [reply sky/serve/replica_managers.py:276]: let me remove them

**Comment by MaoZiming** [comment]
Thanks for the comments! Fixed

File: sky/serve/replica_managers.py [hunk 12]
```diff
@@ -472,6 +482,7 @@ def _sync_down_logs():
         info = serve_state.get_replica_info_from_id(self.service_name,
                                                     replica_id)
         assert info is not None
+        logger.info(f'preempted: {info.status_property.preempted}')
```
Comment (cblmemo) [inline sky/serve/replica_managers.py:485]: Remove or make it more informative? (e.g. add replica id)

File: sky/serve/replica_managers.py [hunk 13]
```diff
@@ -673,6 +699,36 @@ def _probe_all_replicas(self) -> None:
...
+
+                    if cluster_status != status_lib.ClusterStatus.UP:
+                        # The cluster is (partially) preempted.
+                        # It can be down, INIT or STOPPED, based on the
+                        # interruption behavior of the cloud.
+                        # Spot recovery is needed.
+                        cluster_status_str = (
+                            '' if cluster_status is None else
+                            f' (status: {cluster_status.value})')
+                        logger.info(f'Cluster {info.cluster_name} '
```
Comment (cblmemo) [inline sky/serve/replica_managers.py:726]:
```suggestion
                        logger.info(f'Replica {info.replica_id} '
```

File: tests/test_smoke.py [hunk 14]
```diff
@@ -2734,6 +2734,62 @@ def generate_llm_test_command(prompt: str, expected_output: str) -> str:
...
+                     f'"(labels.ray-cluster-name:{cluster_name})" '
+                     f'--zones={zone} --format="value(name)"')
+        return (f'gcloud compute instances delete --zone={zone}'
+                f' --quiet $({query_cmd})')
+
+    test = Test(
+        f'test-skyserve-spot-recovery-gcp',
+        [
+            f'sky serve up -n {name} -y tests/skyserve/spot/spot_recovery.yaml',
+            _SERVE_WAIT_UNTIL_READY.format(name=name, replica_num=1),
```
Comment (cblmemo) [inline tests/test_smoke.py:2756]:
```suggestion
            _SERVE_WAIT_UNTIL_READY.format(name=name, replica_num=2),
```
iiuc the `spot_recovery.yaml` is using 2 replicas?

File: tests/skyserve/spot/spot_recovery.yaml [hunk 15]
```diff
@@ -0,0 +1,16 @@
+resources:
+  cloud: gcp
+  cpus: 2+
+  use_spot: true
```
Comment (cblmemo) [inline tests/skyserve/spot/spot_recovery.yaml:4]:
```suggestion
  zone: us-central1-a
  use_spot: True
```
Add zone for the smoke test gcloud delete cmd; also nit to keep align with non_recovery.yaml

File: tests/skyserve/spot/nonrecovery.yaml
Comment (cblmemo) [inline tests/skyserve/spot/nonrecovery.yaml:1]: nit: filename `nonrecovery.yaml` -> `spot_user_bug` / `spot_non_recovery`?

**Review by cblmemo — APPROVED** [review]
Awesome!! Left several nits and after all smoke tests passed, it should be good to go! 🙌🏻
Pls run all smoke tests to ensure we don't break anything ;)

File: tests/skyserve/spot/nonrecovery.yaml
Comment (cblmemo) [reply tests/skyserve/spot/nonrecovery.yaml:1]: or probably rename the `spot_recovery.yaml` to `recovery.yaml` since they are already under a dir `spot`

**Pull request merged**